### PR TITLE
Constraints docs [WIP]

### LINKED ISSUE #469: Assembly docs
Add a docs chapter about using assemblies and constrains.

### PR DESCRIPTION
**Pull request by adam-urbanczyk**
This PR will resolve #469 .

### CODE REVIEW

**Comment by adam-urbanczyk** [comment]
@jmwright would you know why this happens:
```
/home/docs/checkouts/readthedocs.org/user_builds/cadquery/checkouts/524/doc/index.rst:33: WARNING: toctree contains reference to nonexisting document 'assy'
```

**Comment by jmwright** [comment]
@adam-urbanczyk It looks like you might have deleted the following message, so I'm assuming you figured it out. If not, let me know and I'll reply.

> @jmwright would you know why this happens:

> /home/docs/checkouts/readthedocs.org/user_builds/cadquery/checkouts/524/doc/index.rst:33: WARNING: toctree contains reference to nonexisting document 'assy'

**Comment by adam-urbanczyk** [comment]
It is solved @jmwright , sorry for the noise.

**Comment by adam-urbanczyk** [comment]
OK, I think it is ready for review.

**Comment by jmwright** [comment]
@adam-urbanczyk It looks like you've got a typical black lint check failure.

File: doc/classreference.rst [hunk 1]
```diff
@@ -61,7 +61,6 @@ Selector Classes
     ParallelDirSelector
     DirectionSelector
     DirectionNthSelector
-    RadiusNthSelector
```
Comment (jmwright) [inline doc/classreference.rst:64]: Why was this removed? It was added recently in #504 wasn't it?

File: doc/primer.rst [hunk 2]
```diff
@@ -195,3 +195,74 @@ Resulting in:
...
+        .constrain('part1@faces@>Y','part3@faces@<Y','Axis')
+        .constrain('part1@faces@>Y','part2@faces@<Y','Axis')
+        .constrain('part1?pt1','part3?pt1','Point')
+        .constrain('part1?pt2','part2?pt2','Point')
+        .solve()
+    )
+
+The following constraints are currently implemented:
+
+   :Axis: two normal vectors are anti-coincident or the cosine of the angle between them is equal to the specified value. Can be defined for all entities with consistent normal vector - planar faces, wired and edges.
```
Comment (jmwright) [inline doc/primer.rst:264]: Should "wired" here be "wires"?

**Review by jmwright — APPROVED** [review]
@adam-urbanczyk Thanks! Other than the two comments I've made it looks good.

File: doc/assy.rst [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+Constraints definition
+======================
+
+Then we want to define all the constraints
+
+.. code-block:: python
+  :linenos:
+   
+    # define the elements
+    door = (
```
Comment (marcus7070) [inline doc/assy.rst:161]: I think you've cut and pasted the wrong block, this is the same as the previous section.

**Comment by marcus7070** [comment]
I built this locally and it looks good to me. I think it would be worth introducing the 5 argument version of `Assembly.constrain` since it's the more flexible version, but perhaps that's a job for the next PR.

File: doc/classreference.rst (hunk 1 above)
Comment (adam-urbanczyk) [reply doc/classreference.rst:64]: Some kind of mistake, solved

File: doc/primer.rst (hunk 2 above)
Comment (adam-urbanczyk) [reply doc/primer.rst]: Thanks, solved

File: doc/assy.rst (hunk 3 above)
Comment (adam-urbanczyk) [reply doc/assy.rst]: Good catch, solved

**Comment by adam-urbanczyk** [comment]
Do you see a natural place to introduce it in this doc @marcus7070 ?

File: doc/primer.rst [hunk 4]
```diff
@@ -195,3 +195,74 @@ Resulting in:
...
+        .add(part3, name='part3',color=Color("red"))
+        .constrain('part1@faces@>Z','part3@faces@<Z','Axis')
+        .constrain('part1@faces@>Z','part2@faces@<Z','Axis')
+        .constrain('part1@faces@>Y','part3@faces@<Y','Axis')
+        .constrain('part1@faces@>Y','part2@faces@<Y','Axis')
+        .constrain('part1?pt1','part3?pt1','Point')
+        .constrain('part1?pt2','part2?pt2','Point')
+        .solve()
+    )
+
```
Comment (marcus7070) [inline doc/primer.rst:261]: Around here might be a good spot to introduce the object based constraint method.
```markdown
Should you need to do something unusual that is not possible with these string
based selectors (eg. use a custom selector class), it is possible to pass Shape
objects to the method instead. For example, the above
`.constrain('part1@faces@>Z','part3@faces@<Z','Axis')` is equivalent to
`.constrain('part1',part1.faces('>z').val(),'part3',part3.faces('<Z').val(),'Axis')`.
This method requires a Shape object, so remember to use the `Workplane.val`
method to pass a single Shape and not the whole Workplane object.
```

Comment (adam-urbanczyk) [reply doc/primer.rst:262]: Done, thanks for the suggestion

**Comment by adam-urbanczyk** [comment]
Alright, I think all is resolved now. I propose to merge this today.

**Comment by jmwright** [comment]
+1 to merge
